numpy/lib/mixins.py

```diff
@@ -0,0 +1,167 @@
+"""Mixin classes for custom array types that don't inherit from ndarray."""
+from __future__ import division, absolute_import, print_function
+
+import sys
+
+from numpy.core import umath as um
+
+# None of this module should be exposed in top-level NumPy module.
+__all__ = []
+
+
+def _binary_method(ufunc):
+    def func(self, other):
+        try:
+            if other.__array_ufunc__ is None:
+                return NotImplemented
+        except AttributeError:
+            pass
+        return self.__array_ufunc__(ufunc, '__call__', self, other)
```

Comment:
@mhvk discussed this in part in the "In combination with Python's binary operations" [section of the NEP](https://github.com/charris/numpy/blob/406bbc652424fff332f49b0d2f2e5aedd8191d33/doc/neps/ufunc-overrides.rst#in-combination-with-pythons-binary-operations). I would describe my choice here as a paranoid variant of "Option 2". Let me explain:

- We think both that both Option 1 (call ufunc directly) and Option 2 (call into one's own `__array_ufunc__`) will work, but aren't yet sure which is better in general. There are two reasons to favor Option 2 for the mixin:
    1. It gives a better error message for the `TypeError` that is raised when no operation is available for binary arithmetic (i.e., "unsupported operand type(s) for +" vs something [TBD](https://github.com/charris/numpy/pull/23) about unsupported  `__numpy_ufunc__` for `<ufunc 'add'>`). This argument also applies for `numpy.ndarray`, so possibly we should switch the implementation to match, as @pv noted above.
    2. It still works for implementing binary arithmetic even when backported from NumPy into a third-party library, because it doesn't rely on ufuncs supporting dispatch.
- I call my variant "paranoid" because I don't rely on `__numpy_ufunc__` returning `NotImplemented` if the other argument sets `__numpy_ufunc__ = None`. Yes, a well written version of `__numpy_ufunc__` should probably always do so, because it shouldn't attempt to perform arbitrary operation. But programmers are lazy and/or ignorant of the finer details, and this version ensures that someone cannot inadvertently override binary arithmetic via `__numpy_ufunc__` when doing an operation with a class that opted out of `__numpy_ufunc__`.

(Of course, I'll add this reasoning to comments in code and/or the NEP once we agree on it.)